### PR TITLE
feat(async-component): Add reloadOnVisible

### PR DESCRIPTION
**Pull request by evanpurkhiser**
Setting this flag on a component will cause the component to reload it's data when the window becomes visible and it is mounted.

### CODE REVIEW

**Comment by dcramer** [comment]
noice! will defer to billy on review

**Review by billyvg — COMMENTED** [review]
Hmm I think this would be helpful as a standalone component which `AsyncComponent` would implement. I'm happy to mash approve if you want this merged in first though.

I think the interface should be something like:

```javascript
<VisibilityChange
onShow={}
onHide={}
disabled={}
/>
```

File: src/sentry/static/sentry/app/components/asyncComponent.jsx
Comment (billyvg) [inline src/sentry/static/sentry/app/components/asyncComponent.jsx]: We need a way for subclassed Component to control when to reload, e.g. we don't want to reload every time the user alt-tabs back and forth. I think we probably want another instance variable and by default this doesn't reload data when `reloadOnVisible` is true.

Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/components/asyncComponent.jsx]: Why not just mutate the `reloadOnVisible` instance variable in that case?

**Comment by EvanPurkhiser** [comment]
Fixing the tests now.

If I could ask you to go ahead and mash approve on this for now, at some point in the future we can circle back around on this. The code is super trivial, so not too difficult to port over.

File: src/sentry/static/sentry/app/components/asyncComponent.jsx
Comment (billyvg) [reply src/sentry/static/sentry/app/components/asyncComponent.jsx]: Technically that would work, but it conflates the meaning of `reloadOnVisible` - this feature only works if its value is `true` at the time of mounting, and it's confusing to also mutate throughout its lifecycle.

Comment (billyvg) [reply src/sentry/static/sentry/app/components/asyncComponent.jsx]: Most importantly, by default, it should require implementor to toggle a flag in order to have the reload work (I think a 2nd instance variable makes more sense)

Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/components/asyncComponent.jsx]: > this feature only works if its value is true at the time of mounting

Ah yeah totally true, I overlooked this. I'll add a 2nd toggle flag

**Review by billyvg — APPROVED** [review]

**Pull request merged**
